### ISSUE TITLE
actor_critic_mountaincar.py in chapter 8 doesn't work

### ISSUE THREAD

**Issue by mbreaux** [PacktPublishing/PyTorch-1.x-Reinforcement-Learning-Cookbook#1]
Trying to run actor_critic_mountaincar.py in chapter 8, this doesn't seem to work.  I get results that don't seem to converge.  Typically, the total reward will start at a negative number and hover about +-5 around that:

```
Episode: 189, total reward: -45.45590213491918
Episode: 190, total reward: -44.67584862288027
Episode: 191, total reward: -46.18851439524393
Episode: 192, total reward: -44.87449868080457
Episode: 193, total reward: -42.35902967639213
Episode: 194, total reward: -43.54636324990745
Episode: 195, total reward: -45.358866788658894
Episode: 196, total reward: -44.67404678386565
Episode: 197, total reward: -44.20335494957821
Episode: 198, total reward: -47.168525441340314
Episode: 199, total reward: -43.761677334435674
```
![image](https://user-images.githubusercontent.com/13104047/71649522-880a6a80-2cdd-11ea-82e0-1bd68fa09fef.png)

**Comment by VivienvV** [comment]
I have the same issue! In addition, I get the following warning for each episode:

` UserWarning: Using a target size (torch.Size([])) that is different to the input size (torch.Size([1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.`

Hope this gets solved!

**Comment by CrazyNicolas** [comment]
dqn.py in Chapter 7 also doesn't work, have you guys noticed that?

**Comment by gekator** [comment]
Thanks Qwen-2.5 Max

```
import torch
import gym
import torch.nn as nn
import torch.nn.functional as F
import sklearn.preprocessing
import numpy as np

# Создаем среду
env = gym.make('MountainCarContinuous-v0')

# Модель Actor-Critic
class ActorCriticModel(nn.Module):
    def __init__(self, n_input, n_output, n_hidden):
        super(ActorCriticModel, self).__init__()
        self.fc = nn.Linear(n_input, n_hidden)
        self.mu = nn.Linear(n_hidden, n_output)
        self.sigma = nn.Linear(n_hidden, n_output)
        self.value = nn.Linear(n_hidden, 1)
        self.distribution = torch.distributions.Normal

    def forward(self, x):
        x = F.relu(self.fc(x))
        mu = 2 * torch.tanh(self.mu(x))  # Ограничиваем mu в диапазоне [-2, 2]
        sigma = F.softplus(self.sigma(x)) + 1e-3  # Добавляем малое значение для стабильности
        dist = self.distribution(mu.view(1, ), sigma.view(1, ))
        value = self.value(x)
        return dist, value


class PolicyNetwork():
    def __init__(self, n_state, n_action, n_hidden, lr=0.001):
        self.model = ActorCriticModel(n_state, n_action, n_hidden)
        self.optimizer = torch.optim.Adam(self.model.parameters(), lr)

    def update(self, returns, log_probs, state_values):
        """
        Обновление весов модели на основе потерь
        """
        advantages = returns - torch.cat(state_values).squeeze()
        policy_loss = (-torch.cat(log_probs) * advantages.detach()).mean()
        value_loss = F.smooth_l1_loss(torch.cat(state_values).squeeze(), returns)

        loss = policy_loss + 0.5 * value_loss  # Балансируем потери

        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()

        # Отладочный вывод
        print(f"Policy Loss: {policy_loss.item()}, Value Loss: {value_loss.item()}")

    def predict(self, s):
        """
        Предсказание распределения и значения состояния
        """
        self.model.training = False
        return self.model(torch.Tensor(s))

    def get_action(self, s):
        """
        Получение действия и его логарифма вероятности
        """
        dist, state_value = self.predict(s)
        action = dist.sample().clamp(-2, 2)  # Ограничиваем действие в диапазоне [-2, 2]
        log_prob = dist.log_prob(action[0]).clamp(-10, 10)  # Ограничиваем log_prob
        return action, log_prob, state_value


def scale_state(state):
    """
    Нормализация состояния
    """
    scaled = scaler.transform([state])
    return scaled[0]


def actor_critic(env, estimator, n_episode, gamma=0.99, render=False):
    """
    Алгоритм Actor-Critic для непрерывных действий
    """
    for episode in range(n_episode):
        log_probs = []
        rewards = []
        state_values = []
        state = env.reset()

        while True:
            state = scale_state(state)
            action, log_prob, state_value = estimator.get_action(state)
            action = np.clip(action.item(), env.action_space.low[0], env.action_space.high[0])  # Ограничиваем действие
            next_state, reward, is_done, _ = env.step([action])
            
            if render == True:
                env.render()
                
            total_reward_episode[episode] += reward
            log_probs.append(log_prob)
            state_values.append(state_value)
            rewards.append(reward)

            if is_done:
                returns = []
                Gt = 0
                pw = 0

                # Вычисляем дисконтированные награды
                for reward in rewards[::-1]:
                    Gt += gamma ** pw * reward
                    pw += 1
                    returns.append(Gt)

                returns = torch.tensor(returns[::-1])
                returns = (returns - returns.mean()) / (returns.std() + 1e-9)  # Стандартизируем награды

                # Обновляем модель
                estimator.update(returns, log_probs, state_values)

                print(f"Episode: {episode}, Total Reward: {total_reward_episode[episode]}")
                break

            state = next_state


# Подготовка данных для нормализации
state_space_samples = np.array([env.observation_space.sample() for _ in range(10000)])
scaler = sklearn.preprocessing.StandardScaler()
scaler.fit(state_space_samples)

# Параметры
n_state = env.observation_space.shape[0]
n_action = 1
n_hidden = 128
lr = 0.001
policy_net = PolicyNetwork(n_state, n_action, n_hidden, lr) 

#policy_net.model.load_state_dict(torch.load("MasterNode.pt"))
n_episode = 200
gamma = 0.97
total_reward_episode = [0] * n_episode

# Запуск обучения
actor_critic(env, policy_net, n_episode, gamma, render=True)
```
